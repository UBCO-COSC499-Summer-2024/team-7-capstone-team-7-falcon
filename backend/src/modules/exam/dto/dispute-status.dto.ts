import { IsNotEmpty, IsString } from 'class-validator';
import { DisputeStatusEnum } from '../../../enums/exam-dispute.enum';

export class DisputeStatusDto {
  @IsString()
  @IsNotEmpty()
  status!: DisputeStatusEnum;
}
