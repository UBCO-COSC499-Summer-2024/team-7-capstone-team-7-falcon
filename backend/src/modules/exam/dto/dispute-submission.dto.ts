import { IsString, MinLength } from 'class-validator';

export class DisputeSubmissionDto {
  @IsString()
  @MinLength(1)
  description!: string;
}
