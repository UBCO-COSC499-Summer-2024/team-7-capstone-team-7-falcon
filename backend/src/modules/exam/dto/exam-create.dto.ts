import {
  IsInt,
  IsJSON,
  IsOptional,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';
import { ERROR_MESSAGES } from '../../../common';

export class ExamCreateDto {
  @IsString()
  @MinLength(2)
  @MaxLength(46, { message: ERROR_MESSAGES.examController.examNameTooLong })
  exam_name!: string;

  @IsInt()
  exam_date!: number;

  @IsJSON()
  @IsOptional()
  payload?: JSON;
}
