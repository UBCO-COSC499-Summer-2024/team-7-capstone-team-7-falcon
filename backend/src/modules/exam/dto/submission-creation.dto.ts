import { IsNumber, IsString, Max, Min, ValidateNested } from 'class-validator';
import { ERROR_MESSAGES } from '.././../../common';

export class SubmissionCreationDto {
  @IsNumber(
    { maxDecimalPlaces: 3 },
    { message: ERROR_MESSAGES.submissionController.invalidGradeError },
  )
  @Min(0, { message: ERROR_MESSAGES.submissionController.minGradeError })
  @Max(100, { message: ERROR_MESSAGES.submissionController.maxGradeError })
  score!: number;

  @ValidateNested()
  answers!: any;

  @IsString()
  documentPath!: string;
}
