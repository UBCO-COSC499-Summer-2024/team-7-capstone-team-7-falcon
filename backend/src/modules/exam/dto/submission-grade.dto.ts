import {
  IsBoolean,
  IsNumber,
  IsArray,
  ValidateNested,
  IsObject,
  Min,
  Max,
} from 'class-validator';
import { Type } from 'class-transformer';
import 'reflect-metadata';

class Answer {
  @IsNumber()
  @Min(0)
  question_num!: number;

  @IsArray()
  @IsNumber({}, { each: true })
  expected!: number[];

  @IsArray()
  @IsNumber({}, { each: true })
  answered!: number[];

  @IsNumber()
  @Min(0)
  @Max(100)
  score!: number;
}

class Answers {
  @IsBoolean()
  errorFlag!: boolean;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Answer)
  answer_list!: Answer[];
}

export class SubmissionGradeDto {
  @IsObject()
  @ValidateNested()
  @Type(() => Answers)
  answers!: Answers;
}
