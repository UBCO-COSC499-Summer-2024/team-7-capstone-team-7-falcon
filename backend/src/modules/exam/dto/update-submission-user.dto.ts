import { IsNumber, Min } from 'class-validator';

export class UpdateSubmissionUserDto {
  @IsNumber()
  @Min(1)
  studentId!: number;
}
