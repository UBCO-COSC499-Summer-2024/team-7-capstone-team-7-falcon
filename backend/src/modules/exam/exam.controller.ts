import {
  Body,
  Controller,
  Get,
  HttpStatus,
  Param,
  Patch,
  Post,
  Res,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import { Response } from 'express';
import { Roles } from '../../decorators/roles.decorator';
import { CourseRoleEnum } from '../../enums/user.enum';
import { AuthGuard } from '../../guards/auth.guard';
import { CourseRoleGuard } from '../../guards/course-role.guard';
import { ExamCreateDto } from './dto/exam-create.dto';
import { ExamService } from './exam.service';
import {
  ExamCreationException,
  ExamNotFoundException,
  SubmissionNotFoundException,
} from '../../common/errors';
import { User } from '../../decorators/user.decorator';
import { UserModel } from '../user/entities/user.entity';
import { UpcomingExamsInterface } from '../../common/interfaces';
import { ERROR_MESSAGES } from '../../common';
import { SubmissionGradeDto } from './dto/submission-grade.dto';

@Controller('exam')
export class ExamController {
  /**
   * Constructor of ExamController
   * @param examService {ExamService} instance of ExamService
   */
  constructor(private readonly examService: ExamService) {}

  /**
   * Get exam by id
   * @param res {Response} response object
   * @param eid {number} exam id
   * @returns {Promise<Response>} response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR)
  @Get('/:cid/exam/:eid')
  async getExamById(
    @Res() res: Response,
    @Param('eid', new ValidationPipe()) eid: number,
  ): Promise<Response> {
    try {
      const exam = await this.examService.getExamById(eid);
      return res.status(HttpStatus.OK).send(exam);
    } catch (e) {
      if (e instanceof ExamNotFoundException) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Create an exam for the course
   * @param res {Response} response object
   * @param cid {number} course id
   * @param examData {ExamCreateDto} exam data
   * @returns {Promise<Response>} response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR)
  @Post('/:cid/create')
  async createExam(
    @Res() res: Response,
    @Param('cid', new ValidationPipe()) cid: number,
    @Body(new ValidationPipe()) examData: ExamCreateDto,
  ): Promise<Response> {
    try {
      await this.examService.createExam(cid, examData);
      return res.status(HttpStatus.OK).send({ message: 'ok' });
    } catch (e) {
      if (e instanceof ExamCreationException) {
        return res.status(HttpStatus.BAD_REQUEST).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get all exams for the course
   * @param res {Response} - Response object
   * @param eid {number} - Exam id
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Get('/:cid/:eid/submissions')
  async getExam(
    @Res() res: Response,
    @Param('eid', new ValidationPipe()) eid: number,
  ): Promise<Response> {
    try {
      const exam = await this.examService.getSubmissionsByExamId(eid);
      return res.status(HttpStatus.OK).send(exam);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }

  /**
   * Update grade for the submission
   * @param res {Response} - Response object
   * @param eid {number} - Exam id
   * @param cid {number} - Course id
   * @param sid {number} - Submission id
   * @param body {SubmissionGradeDto} - Submission grade data
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard, CourseRoleGuard)
  @Roles(CourseRoleEnum.PROFESSOR, CourseRoleEnum.TA)
  @Patch('/:eid/course/:cid/submission/:sid/grade')
  async updateGrade(
    @Res() res: Response,
    @Param('eid', new ValidationPipe()) eid: number,
    @Param('cid', new ValidationPipe()) cid: number,
    @Param('sid', new ValidationPipe()) sid: number,
    @Body(new ValidationPipe()) body: SubmissionGradeDto,
  ): Promise<Response> {
    try {
      await this.examService.updateGrade(eid, cid, sid, body.grade);
      return res.status(HttpStatus.OK).send({ message: 'ok' });
    } catch (e) {
      if (e instanceof SubmissionNotFoundException) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: e.message,
        });
      } else {
        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
          message: e.message,
        });
      }
    }
  }

  /**
   * Get upcoming exams by user
   * @param res {Response} - Response object
   * @param user {UserModel} - User object
   * @returns {Promise<Response>} - Response object
   */
  @UseGuards(AuthGuard)
  @Get('/upcoming')
  async getUpcomingExamsByUser(
    @Res() res: Response,
    @User() user: UserModel,
  ): Promise<Response> {
    try {
      const exams: UpcomingExamsInterface[] =
        await this.examService.getUpcomingExamsByUser(user);

      if (exams.length === 0) {
        return res.status(HttpStatus.NOT_FOUND).send({
          message: ERROR_MESSAGES.examController.noUpcomingExamsFound,
        });
      }

      return res.status(HttpStatus.OK).send(exams);
    } catch (e) {
      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
        message: e.message,
      });
    }
  }
}
