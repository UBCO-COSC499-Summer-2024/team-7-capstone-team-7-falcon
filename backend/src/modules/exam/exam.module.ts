import { Module } from '@nestjs/common';
import { ExamController } from './exam.controller';
import { ExamService } from './exam.service';
import { UserService } from '../user/user.service';
import { CourseService } from '../course/course.service';
import { TokenService } from '../token/token.service';

@Module({
  controllers: [ExamController],
  providers: [ExamService, CourseService, UserService, TokenService],
  exports: [ExamService],
})
export class ExamModule {}
