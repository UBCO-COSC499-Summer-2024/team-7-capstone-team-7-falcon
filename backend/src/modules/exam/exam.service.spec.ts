import { Test, TestingModule } from '@nestjs/testing';
import { ExamService } from './exam.service';
import {
  TestConfigModule,
  TestTypeOrmModule,
} from '../../../test/utils/testUtils';
import { ExamCreateDto } from './dto/exam-create.dto';
import { CourseModel } from '../course/entities/course.entity';
import { ExamModel } from './entities/exam.entity';
import { CourseService } from '../course/course.service';
import { UserService } from '../user/user.service';
import { TokenService } from '../token/token.service';

describe('ExamService', () => {
  let examService: ExamService;
  let moduleRef: TestingModule;

  beforeEach(async () => {
    moduleRef = await Test.createTestingModule({
      providers: [ExamService, CourseService, UserService, TokenService],
      imports: [TestTypeOrmModule, TestConfigModule],
    }).compile();

    examService = moduleRef.get<ExamService>(ExamService);
  });

  afterEach(async () => {
    await moduleRef.close();
  });

  describe('createExam', () => {
    it('should create an exam', async () => {
      const currentTime: number =
        parseInt(new Date().getTime().toString()) + 2_000;

      const examDetails: ExamCreateDto = {
        exam_date: currentTime,
        exam_name: 'Test Exam',
      };

      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      await examService.createExam(course.id, examDetails);

      const exam = await ExamModel.findOne({
        where: { course },
      });

      delete exam.created_at;
      delete exam.updated_at;
      delete exam.exam_date;

      expect(exam).toBeDefined();
      expect(exam).toMatchSnapshot();
    });

    it('should throw an error if the course is not found', async () => {
      const currentTime: number =
        parseInt(new Date().getTime().toString()) + 2_000;

      const examDetails: ExamCreateDto = {
        exam_date: currentTime,
        exam_name: 'Test Exam',
      };

      await expect(examService.createExam(1, examDetails)).rejects.toThrow(
        'Course not found',
      );
    });

    it('should throw an error if the exam date is in the past', async () => {
      const currentTime: number =
        parseInt(new Date().getTime().toString()) - 2_000;

      const examDetails: ExamCreateDto = {
        exam_date: currentTime,
        exam_name: 'Test Exam',
      };

      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      await expect(
        examService.createExam(course.id, examDetails),
      ).rejects.toThrow('Exam date must be in the future');
    });
  });
});
