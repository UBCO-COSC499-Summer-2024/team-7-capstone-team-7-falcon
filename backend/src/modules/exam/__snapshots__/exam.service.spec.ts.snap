// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExamService createExam should create an exam 1`] = `
ExamModel {
  "grades_released_at": null,
  "id": 1,
  "name": "Test Exam",
  "questions": {},
}
`;

exports[`ExamService getExamsByCourseId should return course exams from first page 1`] = `
PageDto {
  "data": [
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000009",
      "grades_released_at": null,
      "id": 10,
      "name": "Exam 9",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000008",
      "grades_released_at": null,
      "id": 9,
      "name": "Exam 8",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000007",
      "grades_released_at": null,
      "id": 8,
      "name": "Exam 7",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000006",
      "grades_released_at": null,
      "id": 7,
      "name": "Exam 6",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000005",
      "grades_released_at": null,
      "id": 6,
      "name": "Exam 5",
      "questions": {},
      "updated_at": "1000000000",
    },
  ],
  "meta": PageMetaDto {
    "hasNextPage": true,
    "hasPreviousPage": false,
    "itemCount": 10,
    "page": 1,
    "pageCount": 2,
    "take": 5,
  },
}
`;

exports[`ExamService getExamsByCourseId should return course exams from second page 1`] = `
PageDto {
  "data": [
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000004",
      "grades_released_at": null,
      "id": 5,
      "name": "Exam 4",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000003",
      "grades_released_at": null,
      "id": 4,
      "name": "Exam 3",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000002",
      "grades_released_at": null,
      "id": 3,
      "name": "Exam 2",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000001",
      "grades_released_at": null,
      "id": 2,
      "name": "Exam 1",
      "questions": {},
      "updated_at": "1000000000",
    },
    ExamModel {
      "created_at": "1000000000",
      "exam_date": "1000000000",
      "grades_released_at": null,
      "id": 1,
      "name": "Exam 0",
      "questions": {},
      "updated_at": "1000000000",
    },
  ],
  "meta": PageMetaDto {
    "hasNextPage": false,
    "hasPreviousPage": true,
    "itemCount": 10,
    "page": 2,
    "pageCount": 2,
    "take": 5,
  },
}
`;

exports[`ExamService getSubmissionsByExamId should return submissions by exam id 1`] = `
[
  {
    "created_at": "1000000000",
    "id": 10,
    "score": 9,
    "student": {
      "id": 10,
      "student_id": "9",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 10,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 9,
    "score": 8,
    "student": {
      "id": 9,
      "student_id": "8",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 9,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 8,
    "score": 7,
    "student": {
      "id": 8,
      "student_id": "7",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 8,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 7,
    "score": 6,
    "student": {
      "id": 7,
      "student_id": "6",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 7,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 6,
    "score": 5,
    "student": {
      "id": 6,
      "student_id": "5",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 6,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 5,
    "score": 4,
    "student": {
      "id": 5,
      "student_id": "4",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 5,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 4,
    "score": 3,
    "student": {
      "id": 4,
      "student_id": "3",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 4,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 3,
    "score": 2,
    "student": {
      "id": 3,
      "student_id": "2",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 3,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 2,
    "score": 1,
    "student": {
      "id": 2,
      "student_id": "1",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 2,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
  {
    "created_at": "1000000000",
    "id": 1,
    "score": 0,
    "student": {
      "id": 1,
      "student_id": "0",
      "user": {
        "avatar_url": null,
        "first_name": "John",
        "id": 1,
        "last_name": "Doe",
      },
    },
    "updated_at": "1000000000",
  },
]
`;

exports[`ExamService getUpcomingExamsByCourseUser should return upcoming exams for user 1`] = `
[
  {
    "courseCode": "CS101",
    "courseId": 1,
    "courseName": "Introduction to Computer Science",
    "exams": [
      {
        "id": 1,
        "name": "Exam 0",
      },
      {
        "id": 2,
        "name": "Exam 1",
      },
      {
        "id": 3,
        "name": "Exam 2",
      },
      {
        "id": 4,
        "name": "Exam 3",
      },
      {
        "id": 5,
        "name": "Exam 4",
      },
      {
        "id": 6,
        "name": "Exam 5",
      },
      {
        "id": 7,
        "name": "Exam 6",
      },
      {
        "id": 8,
        "name": "Exam 7",
      },
      {
        "id": 9,
        "name": "Exam 8",
      },
      {
        "id": 10,
        "name": "Exam 9",
      },
    ],
  },
  {
    "courseCode": "CS102",
    "courseId": 2,
    "courseName": "Introduction to Computer Science II",
    "exams": [
      {
        "id": 11,
        "name": "Exam 0",
      },
      {
        "id": 12,
        "name": "Exam 1",
      },
      {
        "id": 13,
        "name": "Exam 2",
      },
      {
        "id": 14,
        "name": "Exam 3",
      },
      {
        "id": 15,
        "name": "Exam 4",
      },
      {
        "id": 16,
        "name": "Exam 5",
      },
      {
        "id": 17,
        "name": "Exam 6",
      },
      {
        "id": 18,
        "name": "Exam 7",
      },
      {
        "id": 19,
        "name": "Exam 8",
      },
      {
        "id": 20,
        "name": "Exam 9",
      },
    ],
  },
]
`;
